### PR TITLE
add closable protocol

### PR DESCRIPTION
**Pull request by tomerd**
motivation: marker protocol for entitites that manage resources and need to be closed to to cleanly free such resources

changes:
* define Closable protocol
* conform WritableByteStrea to Closable
* conform PersistenceCache to Closable

### CODE REVIEW

**Comment by tomerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
Hmm, what about the case where you have a shared reference that shouldn't be closed?  e.g. you have a non-owning reference to a file descriptor.  Is this supposed to be specific to SQLite?  At least on Windows, close is not universally applicable - some handles are released, some or closed, etc.

File: Sources/TSCUtility/PersistenceCache.swift [hunk 1]
```diff
@@ -39,6 +39,10 @@ public final class SQLiteBackedPersistentCache: PersistentCacheProtocol {
     deinit {
         try? db.close()
```
Comment (abertelrud) [inline Sources/TSCUtility/PersistenceCache.swift:40]: I wonder if this should call just `close()` now?  It's a final class, so I guess it doesn't matter, but seems like a more correct pattern in general.

Comment (abertelrud) [reply Sources/TSCUtility/PersistenceCache.swift:40]: Arguably it should always have called `self.close()` — it's not a consequence of this particular diff, just more cleanup.

**Review by abertelrud — APPROVED** [review]
Looks great to me.

File: Sources/TSCUtility/PersistenceCache.swift (hunk 1 above)
Comment (tomerd) [reply Sources/TSCUtility/PersistenceCache.swift:40]: good point

**Comment by tomerd** [comment]
> Hmm, what about the case where you have a shared reference that shouldn't be closed? e.g. you have a non-owning reference to a file descriptor. Is this supposed to be specific to SQLite? At least on Windows, close is not universally applicable - some handles are released, some or closed, etc.

Are you referring the name "close" semantically compared to "release"? In the two cases where this applied in this PR the original method was already named `close` so the semantics is preserved.

Is there anything different you want to see here?

**Comment by tomerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
Well, I would like to see a bit more documentation on the intent of the protocol and the intended usage.  If the usage is meant to be localised to these two sites, then that name should be fine.  I'm just thinking about what would happen if the protocol ends up being adopted more widely as some platforms don't follow the "everything is a fd" model, and this can raise unnecessary confusion.  The current name is pretty broad and so the documentation would be helpful for a casual reader.

**Comment by tomerd** [comment]
@swift-ci please test

**Comment by tomerd** [comment]
> Well, I would like to see a bit more documentation on the intent of the protocol and the intended usage. If the usage is meant to be localised to these two sites, then that name should be fine. I'm just thinking about what would happen if the protocol ends up being adopted more widely as some platforms don't follow the "everything is a fd" model, and this can raise unnecessary confusion. The current name is pretty broad and so the documentation would be helpful for a casual reader.

got it, thanks for pointing it out. added a comment on intent of API

**Pull request merged**